iib/web/models.py

```diff
@@ -377,6 +335,101 @@ def _state_to_json(state):
 
         return rv
 
+
+def get_request_query_options(verbose=False):
+    """
+    Get the query options for a SQLAlchemy query for one or more requests to output as JSON.
+
+    This will add the joins ahead of time on relationships that are accessed in the ``to_json``
+    methods to avoid individual select statements when the relationships are accessed.
+
+    :param bool verbose: if the request relationships should be loaded for verbose JSON output
+    :return: a list of SQLAlchemy query options
+    :rtype: list
+    """
+    # Tell SQLAlchemy to join on the relationships that are part of the JSON to avoid
+    # additional SQL queries
+    query_options = [
+        joinedload(Request.user),
+        joinedload(RequestAdd.binary_image),
+        joinedload(RequestAdd.binary_image_resolved),
+        joinedload(RequestAdd.bundles),
+        joinedload(RequestAdd.from_index),
+        joinedload(RequestAdd.from_index_resolved),
+        joinedload(RequestAdd.index_image),
+        joinedload(RequestRm.binary_image),
+        joinedload(RequestRm.binary_image_resolved),
+        joinedload(RequestRm.from_index),
+        joinedload(RequestRm.from_index_resolved),
+        joinedload(RequestRm.index_image),
+        joinedload(RequestRm.operators),
+    ]
+    if verbose:
+        query_options.append(joinedload(Request.states))
+    else:
+        query_options.append(joinedload(Request.state))
+
+    return query_options
+
+
+class RequestIndexImageMixin:
+    """
+    A class for shared functionality between index image requests.
+
+    This class uses the Mixin pattern as defined in:
+    https://docs.sqlalchemy.org/en/13/orm/extensions/declarative/mixins.html
+    """
+
+    @declared_attr
+    def binary_image_id(cls):
+        """Return the ID of the image that the opm binary comes from."""
+        return db.Column(db.Integer, db.ForeignKey('image.id'), nullable=False)
+
+    @declared_attr
+    def binary_image_resolved_id(cls):
+        """Return the ID of the resolved image that the opm binary comes from."""
+        return db.Column(db.Integer, db.ForeignKey('image.id'))
+
+    @declared_attr
+    def binary_image(cls):
+        """Return the relationship to the image that the opm binary comes from."""
+        return db.relationship('Image', foreign_keys=[cls.binary_image_id], uselist=False)
+
+    @declared_attr
+    def binary_image_resolved(cls):
+        """Return the relationship to the resolved image that the opm binary comes from."""
+        return db.relationship('Image', foreign_keys=[cls.binary_image_resolved_id], uselist=False)
+
+    @declared_attr
+    def from_index_id(cls):
+        """Return the ID of the index image to base the request from."""
+        return db.Column(db.Integer, db.ForeignKey('image.id'))
```

Comment:
Could you make this `nullable=False` on the `RequestRm` table? It only allowed this to be nullable at the database level for those requests because the table was shared with the add requests.

Reply:
Done. I also modified `test_models.py` to better represent this.